File: Wildcard Matching.cpp
class Solution {
public:
    bool isMatch(const char *s, const char *p) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (!s && !p) return true;
        const char *starP = NULL, *starS = NULL;
        
        while (*s != '\0') {
            if (*p == '?' || *s == *p) {
                s++;
                p++;
            } else if (*p == '*') {
                while (*p == '*') p++;
                if (*p == '\0') return true;
                starP = p;
                starS = s;
                
            } else if ((*p == '\0' || *p != *s) && starP) {
                p = starP;
                s = ++starS;
            } else {
                return false;
            }
            
        }
        
        while (*p != '\0') {
            if (*p++ != '*') return false;
        }
        return true;
    }
};

File: Wildcard Matching.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        if (string.IsNullOrEmpty(s) && string.IsNullOrEmpty(p)) {
            return true;
        }
        
        var i = 0;
        var j = 0;
        var starS = -1;
        var starP = -1;
        while (i < s.Length) {
            if (j < p.Length && (p[j] == '?' || s[i] == p[j])) {
                i++;
                j++;
            } else if (j < p.Length && p[j] == '*') {
                // skip continuous '*'
                while (j < p.Length && p[j] == '*') {
                    j++;
                }
                
                if (j == p.Length) {
                    return true;
                }
                
                // record current star's j and i index.
                starP = j;
                starS = i;
            } else {
                if (starS == -1) {
                    return false;
                }
                
                j = starP; // back to the star's index.
                i = ++starS;
            }
        }
        
        while (j < p.Length && p[j] == '*') {
            j++;
        }
        return j == p.Length;
    }
}
